# Project Title - HR Employee Attrition Prediction

#### Description-

In this project, we need to predict whether a give employee will leave the company or want to work continue with it. This dataset consists of categorical and numerical data. Attrition is the target column in the dataset and rest other can be features to train any Machine Learning model. Let's analyze the dataset and build an appropriate Machine Learning model.

We will create a model with the following steps:
1. Import the relevant packages
2. Download and explore the dataset
3. Perform EDA, Apply dataset for preprocessing
4. Predict the target columns

### Importing required libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import RobustScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

### Loding Dataset.

In [2]:
df = pd.read_csv("HR_Employee_Attrition-1.csv")

In [3]:
df.head(10) #Print the first 10 rows

EmployeeNumber Attrition  Age     BusinessTravel  DailyRate  \
0               1       Yes   41      Travel_Rarely       1102   
1               2        No   49  Travel_Frequently        279   
2               3       Yes   37      Travel_Rarely       1373   
3               4        No   33  Travel_Frequently       1392   
4               5        No   27      Travel_Rarely        591   
5               6        No   32  Travel_Frequently       1005   
6               7        No   59      Travel_Rarely       1324   
7               8        No   30      Travel_Rarely       1358   
8               9        No   38  Travel_Frequently        216   
9              10        No   36      Travel_Rarely       1299   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   
5  Research & Development                 2          2  Life Sciences   
6  Research & Development                 3          3        Medical   
7  Research & Development                24          1  Life Sciences   
8  Research & Development                23          3  Life Sciences   
9  Research & Development                27          3        Medical   

   EmployeeCount  ...  RelationshipSatisfaction StandardHours  \
0              1  ...                         1            80   
1              1  ...                         4            80   
2              1  ...                         2            80   
3              1  ...                         3            80   
4              1  ...                         4            80   
5              1  ...                         3            80   
6              1  ...                         1            80   
7              1  ...                         2            80   
8              1  ...                         2            80   
9              1  ...                         2            80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                  8                      0               1   
1                 1                 10                      3               3   
2                 0                  7                      3               3   
3                 0                  8                      3               3   
4                 1                  6                      3               3   
5                 0                  8                      2               2   
6                 3                 12                      3               2   
7                 1                  1                      2               3   
8                 0                 10                      2               3   
9                 2                 17                      3               2   

   YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                  4                        0   
1              10                  7                        1   
2               0                  0                        0   
3               8                  7                        3   
4               2                  2                        2   
5               7                  7                        3   
6               1                  0                        0   
7               1                  0                        0   
8               9                  7                        1   
9               7                  7                        7   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                 

In [4]:
#Get the number of rows and number of columns in the data
df.shape

(2940, 35)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2940 entries, 0 to 2939
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   EmployeeNumber            2940 non-null   int64 
 1   Attrition                 2940 non-null   object
 2   Age                       2940 non-null   int64 
 3   BusinessTravel            2940 non-null   object
 4   DailyRate                 2940 non-null   int64 
 5   Department                2940 non-null   object
 6   DistanceFromHome          2940 non-null   int64 
 7   Education                 2940 non-null   int64 
 8   EducationField            2940 non-null   object
 9   EmployeeCount             2940 non-null   int64 
 10  EnvironmentSatisfaction   2940 non-null   int64 
 11  Gender                    2940 non-null   object
 12  HourlyRate                2940 non-null   int64 
 13  JobInvolvement            2940 non-null   int64 
 14  JobLevel                

In [6]:
#Get the column data types
df.dtypes

EmployeeNumber               int64
Attrition                   object
Age                          int64
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

## Data Description

In [7]:
# View some basic statistical details like percentile, mean, standard deviation etc.
df.describe()

EmployeeNumber          Age    DailyRate  DistanceFromHome  \
count     2940.000000  2940.000000  2940.000000       2940.000000   
mean      1470.500000    36.923810   802.485714          9.192517   
std        848.849221     9.133819   403.440447          8.105485   
min          1.000000    18.000000   102.000000          1.000000   
25%        735.750000    30.000000   465.000000          2.000000   
50%       1470.500000    36.000000   802.000000          7.000000   
75%       2205.250000    43.000000  1157.000000         14.000000   
max       2940.000000    60.000000  1499.000000         29.000000   

         Education  EmployeeCount  EnvironmentSatisfaction   HourlyRate  \
count  2940.000000         2940.0              2940.000000  2940.000000   
mean      2.912925            1.0                 2.721769    65.891156   
std       1.023991            0.0                 1.092896    20.325969   
min       1.000000            1.0                 1.000000    30.000000   
25%       2.000000            1.0                 2.000000    48.000000   
50%       3.000000            1.0                 3.000000    66.000000   
75%       4.000000            1.0                 4.000000    84.000000   
max       5.000000            1.0                 4.000000   100.000000   

       JobInvolvement     JobLevel  ...  RelationshipSatisfaction  \
count     2940.000000  2940.000000  ...               2940.000000   
mean         2.729932     2.063946  ...                  2.712245   
std          0.711440     1.106752  ...                  1.081025   
min          1.000000     1.000000  ...                  1.000000   
25%          2.000000     1.000000  ...                  2.000000   
50%          3.000000     2.000000  ...                  3.000000   
75%          3.000000     3.000000  ...                  4.000000   
max          4.000000     5.000000  ...                  4.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         2940.0       2940.000000        2940.000000   
mean            80.0          0.793878          11.279592   
std              0.0          0.851932           7.779458   
min             80.0          0.000000           0.000000   
25%             80.0          0.000000           6.000000   
50%             80.0          1.000000          10.000000   
75%             80.0          1.000000          15.000000   
max             80.0          3.000000          40.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count            2940.000000      2940.000000     2940.000000   
mean                2.799320         2.761224        7.008163   
std                 1.289051         0.706356        6.125483   
min                 0.000000         1.000000        0.000000   
25%                 2.000000         2.000000        3.000000   
50%                 3.000000         3.000000        5.000000   
75%                 3.000000         3.000000        9.000000   
max                 6.000000         4.000000       40.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count         2940.000000              2940.000000           2940.000000  
mean             4.229252                 2.187755              4.123129  
std              3.622521                 3.221882              3.567529  
min              0.000000                 0.000000              0.000000  
25%              2.000000                 0.000000              2.000000  
50%              3.000000                 1.000000              3.000000  
75%              7.000000                 3.000000              7.000000  
max             18.000000                15.000000             17.000000  

[8 rows x 26 columns]

## Exploratory Data Analysis & Dataset Cleaning

In [9]:
# Performing Automatic EDA by using Pandas Profiling
from pandas_profiling import ProfileReport
prof = ProfileReport(df)
prof.to_file(output_file='output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [10]:
# Count the empty (NaN, NAN, na) values in each column
df.isnull().sum()

EmployeeNumber              0
Attrition                   0
Age                         0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [11]:
# Check for any missing / null valus in data
df.isnull().values.any()

False

In [12]:
# Get a count of the number of employee attrition, the number of employees that stayed (no) and the number that left (yes)
df['Attrition'].value_counts()

No     2466
Yes     474
Name: Attrition, dtype: int64

In [13]:
# Visualize the number of Employees that stayed and left the Company
sns.countplot(df['Attrition'])

C:\Anaconda\lib\site-packages\seaborn\_decorators.py:36: FutureWarning: Pass the following variable as a keyword arg: x. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.
  warnings.warn(


<AxesSubplot:xlabel='Attrition', ylabel='count'>

In [14]:
# This is the percentage if we just guessed no for attrition
(2466 - 474)/ 2466

0.8077858880778589

In [15]:
# Show the number of employee that left and stayed by age
import matplotlib.pyplot as plt
fig_dims = (14, 5)
fig, ax = plt.subplots(figsize=fig_dims)

#ax = axis
sns.countplot(x='Age', hue='Attrition', data = df, palette="colorblind", ax = ax,  edgecolor=sns.color_palette("dark", n_colors = 1));

In [16]:
# Print all of the data types and their unique values
for column in df.columns:
    if df[column].dtype == object:
        print(str(column) + ':' + str(df[column].unique()))
        print(df[column].value_counts())
        print('___________________________________________________________________')

Attrition:['Yes' 'No']
No     2466
Yes     474
Name: Attrition, dtype: int64
___________________________________________________________________
BusinessTravel:['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
Travel_Rarely        2086
Travel_Frequently     554
Non-Travel            300
Name: BusinessTravel, dtype: int64
___________________________________________________________________
Department:['Sales' 'Research & Development' 'Human Resources']
Research & Development    1922
Sales                      892
Human Resources            126
Name: Department, dtype: int64
___________________________________________________________________
EducationField:['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
Life Sciences       1212
Medical              928
Marketing            318
Technical Degree     264
Other                164
Human Resources       54
Name: EducationField, dtype: int64
__________________________________________________________________

In [17]:
df['StandardHours'].unique()

array([80], dtype=int64)

In [18]:
#Remove unneeded columns

#Remove the column EmployeeNumber
df = df.drop('EmployeeNumber', axis = 1) 

#Remove the column StandardHours
df = df.drop('StandardHours', axis = 1) 

#Remove the column EmployeeCount
df = df.drop('EmployeeCount', axis = 1) 

#Remove the column EmployeeCount
df = df.drop('Over18', axis = 1) 


In [19]:
#Get the Correlation of the columns
df.corr()

Age  DailyRate  DistanceFromHome  Education  \
Age                       1.000000   0.010661         -0.001686   0.208034   
DailyRate                 0.010661   1.000000         -0.004985  -0.016806   
DistanceFromHome         -0.001686  -0.004985          1.000000   0.021042   
Education                 0.208034  -0.016806          0.021042   1.000000   
EnvironmentSatisfaction   0.010146   0.018355         -0.016075  -0.027128   
HourlyRate                0.024287   0.023381          0.031131   0.016775   
JobInvolvement            0.029820   0.046135          0.008783   0.042438   
JobLevel                  0.509604   0.002966          0.005303   0.101589   
JobSatisfaction          -0.004892   0.030571         -0.003669  -0.011296   
MonthlyIncome             0.497855   0.007707         -0.017014   0.094961   
MonthlyRate               0.028051  -0.032182          0.027473  -0.026084   
NumCompaniesWorked        0.299635   0.038153         -0.029251   0.126317   
PercentSalaryHike         0.003634   0.022704          0.040235  -0.011111   
PerformanceRating         0.001904   0.000473          0.027110  -0.024539   
RelationshipSatisfaction  0.053535   0.007846          0.006557  -0.009118   
StockOptionLevel          0.037510   0.042143          0.044872   0.018422   
TotalWorkingYears         0.680381   0.014515          0.004628   0.148280   
TrainingTimesLastYear    -0.019621   0.002453         -0.036942  -0.025100   
WorkLifeBalance          -0.021490  -0.037848         -0.026556   0.009819   
YearsAtCompany            0.311309  -0.034055          0.009508   0.069114   
YearsInCurrentRole        0.212901   0.009932          0.018845   0.060236   
YearsSinceLastPromotion   0.216513  -0.033229          0.010029   0.054254   
YearsWithCurrManager      0.202089  -0.026363          0.014406   0.069065   

                          EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
Age                                      0.010146    0.024287        0.029820   
DailyRate                                0.018355    0.023381        0.046135   
DistanceFromHome                        -0.016075    0.031131        0.008783   
Education                               -0.027128    0.016775        0.042438   
EnvironmentSatisfaction                  1.000000   -0.049857       -0.008278   
HourlyRate                              -0.049857    1.000000        0.042861   
JobInvolvement                          -0.008278    0.042861        1.000000   
JobLevel                                 0.001212   -0.027853       -0.012630   
JobSatisfaction                         -0.006784   -0.071335       -0.021476   
MonthlyIncome                           -0.006259   -0.015794       -0.015271   
MonthlyRate                              0.037600   -0.015297       -0.016322   
NumCompaniesWorked                       0.012594    0.022157        0.015012   
PercentSalaryHike                       -0.031701   -0.009062       -0.017205   
PerformanceRating                       -0.029548   -0.002172       -0.029071   
RelationshipSatisfaction                 0.007665    0.001330        0.034297   
StockOptionLevel                         0.003432    0.050263        0.021523   
TotalWorkingYears                       -0.002693   -0.002334       -0.005533   
TrainingTimesLastYear                   -0.019359   -0.008548       -0.015338   
WorkLifeBalance                          0.027627   -0.004607       -0.014617   
YearsAtCompany                           0.001458   -0.019582       -0.021355   
YearsInCurrentRole                       0.018007   -0.024106        0.008717   
YearsSinceLastPromotion                  0.016194   -0.026716       -0.024184   
YearsWithCurrManager                    -0.004999   -0.020123        0.025976   

                          JobLevel  JobSatisfaction  MonthlyIncome  ...  \
Age                       0.509604        -0.004892       0.497855  ...   
DailyRate                 0.002966         0.030571       0.007

In [20]:
#Visualize the correlation
plt.figure(figsize=(14,14))  #14in by 14in
sns.heatmap(df.corr(), annot=True, fmt='.0%')

<AxesSubplot:>

## Label Encoding

In [21]:
#Transform the data
#Transform non-numerical into numerical col
from sklearn.preprocessing import LabelEncoder

for column in df.columns:
    if (df[column].dtype == np.float64 or df[column].dtype == np.int64):
        continue
    df[column] = LabelEncoder().fit_transform(df[column])
            

In [22]:
df

Attrition  Age  BusinessTravel  DailyRate  Department  DistanceFromHome  \
0             1   41               2       1102           2                 1   
1             0   49               1        279           1                 8   
2             1   37               2       1373           1                 2   
3             0   33               1       1392           1                 3   
4             0   27               2        591           1                 2   
...         ...  ...             ...        ...         ...               ...   
2935          0   36               1        884           1                23   
2936          0   39               2        613           1                 6   
2937          0   27               2        155           1                 4   
2938          0   49               1       1023           2                 2   
2939          0   34               2        628           1                 8   

      Education  EducationField  EnvironmentSatisfaction  Gender  ...  \
0             2               1                        2       0  ...   
1             1               1                        3       1  ...   
2             2               4                        4       1  ...   
3             4               1                        4       0  ...   
4             1               3                        1       1  ...   
...         ...             ...                      ...     ...  ...   
2935          2               3                        3       1  ...   
2936          1               3                        4       1  ...   
2937          3               1                        2       1  ...   
2938          3               3                        4       1  ...   
2939          3               3                        2       1  ...   

      PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0                     3                         1                 0   
1                     4                         4                 1   
2                     3                         2                 0   
3                     3                         3                 0   
4                     3                         4                 1   
...                 ...                       ...               ...   
2935                  3                         3                 1   
2936                  3                         1                 1   
2937                  4                         2                 1   
2938                  3                         4                 0   
2939                  3                         1                 0   

      TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
0                     8                      0                1   
1                    10                      3                3   
2                     7                      3                3   
3                     8                      3                3   
4                     6                      3                3   
...                 ...                    ...              ...   
2935                 17                      3                3   
2936                  9                      5                3   
2937                  6                      0                3   
2938                 17                      3                2   
2939                  6                      3                4   

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0                  6                   4                        0   
1                 10                   7                        1   
2                  0                   0                        0   
3                  8                   7                        3   
4                  2                   2                        2   
...              ...                 ...                      ...   
2935           

In [23]:
#Create a new column at the end of the dataframe that contains the same value 
df['Age_Years'] = df['Age']

In [24]:
#Remove the first column called age 
df = df.drop('Age', axis = 1)

In [25]:
#Show the dataframe
df

Attrition  BusinessTravel  DailyRate  Department  DistanceFromHome  \
0             1               2       1102           2                 1   
1             0               1        279           1                 8   
2             1               2       1373           1                 2   
3             0               1       1392           1                 3   
4             0               2        591           1                 2   
...         ...             ...        ...         ...               ...   
2935          0               1        884           1                23   
2936          0               2        613           1                 6   
2937          0               2        155           1                 4   
2938          0               1       1023           2                 2   
2939          0               2        628           1                 8   

      Education  EducationField  EnvironmentSatisfaction  Gender  HourlyRate  \
0             2               1                        2       0          94   
1             1               1                        3       1          61   
2             2               4                        4       1          92   
3             4               1                        4       0          56   
4             1               3                        1       1          40   
...         ...             ...                      ...     ...         ...   
2935          2               3                        3       1          41   
2936          1               3                        4       1          42   
2937          3               1                        2       1          87   
2938          3               3                        4       1          63   
2939          3               3                        2       1          82   

      ...  RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0     ...                         1                 0                  8   
1     ...                         4                 1                 10   
2     ...                         2                 0                  7   
3     ...                         3                 0                  8   
4     ...                         4                 1                  6   
...   ...                       ...               ...                ...   
2935  ...                         3                 1                 17   
2936  ...                         1                 1                  9   
2937  ...                         2                 1                  6   
2938  ...                         4                 0                 17   
2939  ...                         1                 0                  6   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                         0                1               6   
1                         3                3              10   
2                         3                3               0   
3                         3                3               8   
4                         3                3               2   
...                     ...              ...             ...   
2935                      3                3               5   
2936                      5                3               7   
2937                      0                3               6   
2938                      3                2               9   
2939                      3                4               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
0                      4                        0                     5   
1                      7                        1                     7   
2                      0                        0                     0   
3                      7                        3                     0   
4                      2                        2                     2   


## Dividing Data in Train and Test Split

In [26]:
#Split the data into independent 'X' and dependent 'Y' variables
X = df.iloc[:, 1:df.shape[1]].values
Y = df.iloc[:,0].values
print(df.shape)
print(X.shape)
print(Y.shape)

(2940, 31)
(2940, 30)
(2940,)


In [27]:
#Split the data into 75% training and 25% testing
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)
print(X_train.shape, X_test.shape)

(2058, 30) (882, 30)


In [28]:
X_test

array([[   2,  882,    2, ...,    0,    8,   35],
       [   1,  667,    2, ...,    0,    0,   31],
       [   2,  317,    1, ...,    0,    2,   30],
       ...,
       [   2,  506,    1, ...,    1,    2,   24],
       [   2, 1079,    2, ...,    1,    4,   31],
       [   0, 1229,    2, ...,    0,    7,   36]], dtype=int64)

In [29]:
X_train

array([[  0, 849,   2, ...,   2,   2,  46],
       [  1, 662,   2, ...,   3,   2,  35],
       [  2, 555,   1, ...,   0,   0,  40],
       ...,
       [  2, 328,   1, ...,   0,   2,  50],
       [  0, 457,   1, ...,   1,   3,  22],
       [  1, 807,   1, ...,   0,   0,  43]], dtype=int64)

In [30]:
Y_train

array([0, 1, 0, ..., 0, 0, 1])

In [31]:
Y_test

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,

## Model Building

In [32]:
#Use Random Forest Classification algorithm
from sklearn.ensemble import RandomForestClassifier 
forest = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)

#Get the accuracy on the training data
forest.fit(X_train, Y_train)
forest.predict(X_test)

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,

In [33]:
#Show the confusion matrix and accuracy for  the model on the test data
#Classification accuracy is the ratio of correct predictions to total predictions made.
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(Y_test, forest.predict(X_test))

  
TN = cm[0][0]
TP = cm[1][1]
FN = cm[1][0]
FP = cm[0][1]
  
print(cm)
print('Model Testing Accuracy = "{}!"'.format(  (TP + TN) / (TP + TN + FN + FP)))
print()# Print a new line
score = accuracy_score(Y_test, forest.predict(X_test))
print('Random Forest Classifier Score: ', np.abs(score)*100)

[[737   0]
 [ 32 113]]
Model Testing Accuracy = "0.963718820861678!"

Random Forest Classifier Score:  96.3718820861678


#### Reandom Forest Algorithm is used
#### Accuracy Score is -> 96 %

# Author - Pravin Bhagwat